[com_fields] Encode complex values to JSON

stutteringp0et opened the pull request:
This adjustment allows complex field types (types with more than one input) to store and retrieve data.

I'll be submitting this on the 4.0-dev branch as well

### Summary of Changes

Test value before save, if it's a multidimensional array, json encode it

### Testing Instructions

Hard to say without installing a 3rd party custom field plugin

### Expected result

it should be transparent

### Actual result

transparent on my end

### Documentation Changes Required
none

stutteringp0et committed:
Minor adjustment to field value

This adjustment allows complex field types (types with more than one input) to store and retrieve data.

Quy commented on plugins/system/fields/fields.php:
```diff
@@ -90,6 +90,11 @@ public function onContentAfterSave($context, $item, $isNew, $data = array())
 			$value = key_exists($field->name, $fieldsData) ? $fieldsData[$field->name] : null;
 
+			// JSON encode value for complex fields
+			if(is_array($value)) {
```
Please fix coding style as follows:
```
if (is_array($value))
{
```

mbabker commented:
> I'll be submitting this on the 4.0-dev branch as well

No need, we're still branch merging things up so anything going into 3.x eventually lands on 4.0.

stutteringp0et committed:
coding style compliance

Quy commented on plugins/system/fields/fields.php:
```diff
@@ -90,6 +90,12 @@ public function onContentAfterSave($context, $item, $isNew, $data = array())
+			{
+				$value = json_encode($value);
+			}
+			
```
Also please remove tabs on line 98.

stutteringp0et replied:
done

stutteringp0et committed:
coding style compliance

stutteringp0et committed:
more rigorous test

single dimension arrays are already handled, so this alteration changes the behavior to limit the effect to multi-dimensional arrays.

rdeutz committed:
Update fields.php

stutteringp0et commented:
Anyone wanting to test this - I have made a field plugin that creates a multi-dimensional value, but I would rather not post it publicly.

JeroenMoolenschot commented:
Hi stutteringp0et,

I would like to test it, but can you give me the testing-instructions<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19025">issues.joomla.org/tracker/joomla-cms/19025</a>.</sub>

stutteringp0et commented:
@Joomill - I'll send you a message through the email address on your website.  I haven't released the custom plugin and I'd like to keep it under wraps until I can release it.  I had a look at your site and extensions, so I think you'll understand why.

edit - email sent

JeroenMoolenschot commented:
I have tested this item :white_check_mark: successfully on 13533fb3f93f95fdd9ec3824e1543ab2a1dd0445<br /><br />Subform is saved as: {"subform0":{"example_text":"value","example_textarea"...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19025">issues.joomla.org/tracker/joomla-cms/19025</a>.</sub>

stutteringp0et commented:
Thanks for taking the time @Joomill

Quy commented:
@stutteringp0et Please send plugin to test to qubertman@yahoo.com. I will delete this reply immediately to not expose my email address. Thanks.

Quy commented:
I have tested this item :white_check_mark: successfully on 13533fb3f93f95fdd9ec3824e1543ab2a1dd0445<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19025">issues.joomla.org/tracker/joomla-cms/19025</a>.</sub>

Quy commented:
Ready to Commit after two successful tests.

stutteringp0et commented:
Yay!
